java-util/src/main/java/io/druid/java/util/common/io/NativeIO.java

```diff
@@ -0,0 +1,168 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.java.util.common.io;
+
+import java.io.File;
+import java.io.FileDescriptor;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.RandomAccessFile;
+import java.lang.reflect.Field;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import com.sun.jna.LastErrorException;
+import com.sun.jna.Native;
+import com.sun.jna.Platform;
+
+/**
+ * Native I/O operations in order to minimize cache impact.
+ */
+public class NativeIO
```

Comment:
am wondering if it is possible to use a well-known library to do this? am not an expert on the matter, but usually, this kind of very low-level IO operation is hard to get right...

Reply:
This is the canonical way of implementing it:

1. https://github.com/apache/bookkeeper/blob/master/bookkeeper-server/src/main/java/org/apache/bookkeeper/util/NativeIO.java
2. https://github.com/Netflix/sstable-adaptor/blob/master/sstable-adaptor-cassandra/src/main/java/org/apache/cassandra/utils/CLibrary.java

Please note that skipping the cache seems platform dependent, i.e. Linux only (fadvise).

Reply:
I think we are beyond Apache Bookkeeper especially given the latest changes. Also the core logic of the copy function is not in Apache Bookkeeper. I used Apache Cassandra initially as inspiration, then Bookkeeper and some stack overflow.